overlord/exportstate/manifest.go

```diff
@@ -0,0 +1,211 @@
+	}
+	err := os.Symlink(s.Target, pathName)
+	if err != nil && os.IsExist(err) {
+		if actualTarget, _ := os.Readlink(pathName); actualTarget == s.Target {
+			err = nil
+		}
+	}
+	return err
+}
+
+// Remove removes the symbolic link and prunes any directories.
+func (s *SymlinkExport) Remove() error {
+	if err := os.Remove(s.PathName()); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	// XXX: or iterate upwards until we reach ExportDir
+	os.Remove(filepath.Join(ExportDir, s.SnapName, s.ExportedVersion, s.ExportSet))
+	os.Remove(filepath.Join(ExportDir, s.SnapName, s.ExportedVersion))
+	os.Remove(filepath.Join(ExportDir, s.SnapName))
+	return nil
+}
+
+// exportSetSymlinks returns symlink exports for given snap and export entries.
+func exportSetSymlinks(snapName, snapRev, exportSetName string, entries []*exportEntry) []SymlinkExport {
+	symlinks := make([]SymlinkExport, 0, len(entries))
+	for _, entry := range entries {
+		var target string
+		if entry.IsExportedPathValidInHostMountNS {
+			target = entry.PathInHostMountNS
+		} else {
+			target = entry.PathInSnapMountNS
+		}
+		symlinks = append(symlinks, SymlinkExport{
+			SnapName:        snapName,
+			ExportedVersion: snapRev,
+			ExportSet:       exportSetName,
+			Name:            entry.PathInExportSet,
+			Target:          target,
+		})
+	}
+	return symlinks
+}
+
+// exportEntry describes a single file placed in a specific export set.
+//
+// The original file is described by two paths. PathInHostMountNS is is valid
+// in the host mount namespace while PathInSnapMountNS is valid in the per-snap
+// mount namespace. The original file is exported by creating a symbolic link
+// under PathInExportSet, which is relative to the export set that the entry
+// belongs to, to one of the two other paths.
+//
+// If IsExportedPathValidInHostMountNS returns true then PathInHostMountNS is
+// used, if it returns false PathInSnapMountNS is used instead.
+//
+// This distinction enables exporting files that are consumed by either other
+// snaps or by the classic system.
+type exportEntry struct {
+	PathInExportSet   string
+	PathInHostMountNS string
+	PathInSnapMountNS string
+
+	IsExportedPathValidInHostMountNS bool
+}
+
+// newExportedHostFile returns an entry describing a file from the classic file system.
+func newExportedHostFile(pathOnHost, pathInExportSet string) *exportEntry {
+	// TODO: consider using this to describe nvidia libraries from the host.
+	return &exportEntry{
+		PathInHostMountNS: pathOnHost,
+		PathInSnapMountNS: filepath.Join("/var/lib/snapd/hostfs", pathOnHost),
+		PathInExportSet:   pathInExportSet,
+	}
+}
+
+// newExportedSnapFile returns an entry describing a file stored inside a snap.
+func newExportedSnapFile(snap *snap.Info, pathInSnap, pathInExportSet string) *exportEntry {
+	return &exportEntry{
+		PathInHostMountNS: filepath.Join(snap.MountDir(), pathInSnap),
+		PathInSnapMountNS: filepath.Join("/snap", snap.InstanceName(), snap.Revision.String(), pathInSnap),
```

Comment:
the two paths have the same value when not running with a special RootDir, right?

Reply:
Not exactly, some distributions will use `/var/lib/snapd/snap` as the base of MountDir